File: 03 - OhadsCarRental Website/Scripts/CustomValidations/IdentityNumberValidation.js
/*===================================================================================
  IdentityNumberValidation.js ==> Represents a Client Side Validation:
  --------------------------------------------------------------------
  Validates if the identity number is valid, by an algorithm to check IDs.
===================================================================================*/

$.validator.unobtrusive.adapters.add("idnumber", ["other"], function (options) {
    options.rules["idnumber"] = "#" + options.params.other;
    options.messages["idnumber"] = options.message; // The error message
});

$.validator.addMethod("idnumber", function (value, element, params) {

    // Checks if the identity number is valid:
    if (!ValidateID(value))
        return false; // Invalid IdentityNumber!

    return true; // Valid IdentityNumber!
});

// Validates the ID number!
function ValidateID(str) {
    //INPUT VALIDATION

    // Just in case -> convert to string
    var IDnum = String(str);

    // Validate correct input:
    if ((IDnum.length > 9) || (IDnum.length < 5))
        return false; // Invalid id
    if (isNaN(IDnum))
        return false; // Invalid id

    // If the number is too short, add leading 0000:
    if (IDnum.length < 9) {
        while (IDnum.length < 9) {
            IDnum = '0' + IDnum;
        }
    }

    // CHECKS THE ID NUMBER:
    var mone = 0, incNum;
    for (var i = 0; i < 9; i++) {
        incNum = Number(IDnum.charAt(i));
        incNum *= (i % 2) + 1;
        if (incNum > 9)
            incNum -= 9;
        mone += incNum;
    }
    if (mone % 10 == 0)
        return true; // Valid id
    else
        return false; // Invalid id
}

File: 01 - Data Access/Validations/IdentityNumberValidation.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Web.Mvc;

namespace OhadsCarRental.Validations
{
    /// <summary>
    /// Validates if the identity number is valid, by an algorithm to check IDs.
    /// </summary>
    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = true, Inherited = false)]
    public sealed class IdentityNumberValidationAttribute : ValidationAttribute, IClientValidatable
    {

        /// <summary>
        /// Validates if identity number is valid!
        /// </summary>
        /// <param name="idNum">The identity number to validate.</param>
        /// <returns>True of False, if ID is valid or not.</returns>
        private bool ValidateID(string idNum)
        {
            // Validates correct input:
            if (!System.Text.RegularExpressions.Regex.IsMatch(idNum, @"^\d{5,9}$"))
                return false; // Identity Number is invalid!

            // If the number is too short, adds leading 0000:
            if (idNum.Length < 9)
            {
                while (idNum.Length < 9)
                {
                    idNum = '0' + idNum;
                }
            }

            // CHECKS THE ID NUMBER:
            int counter = 0;
            int incNum;
            for (int i = 0; i < 9; i++)
            {
                incNum = Convert.ToInt32(idNum[i].ToString());
                incNum *= (i % 2) + 1;
                if (incNum > 9)
                    incNum -= 9;
                counter += incNum;
            }
            if (counter % 10 == 0)
                return true; // Identity Number is valid!
            else
                return false; // Identity Number is invalid!
        }

        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
        {
            if (value != null)
            {
                if (!ValidateID(value.ToString())) // Checks the ID if not valid...
                {
                    // If so, the Identity Number is invalid!
                    var errorMessage = FormatErrorMessage(validationContext.DisplayName);
                    return new ValidationResult(errorMessage);
                }
            }

            return ValidationResult.Success;
        }

        // For the client side validation:
        public IEnumerable<ModelClientValidationRule> GetClientValidationRules(ModelMetadata metadata, ControllerContext context)
        {
            var rule = new ModelClientValidationRule
            {
                ValidationType = "idnumber", // This is the name jQuery adapter will use!
                ErrorMessage = ErrorMessage // This is the error message the client shows!
            };

            yield return rule;
        }
    }
}
